posix/syscall.c

```diff
@@ -0,0 +1,598 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/corelibc/dirent.h>
+#include <openenclave/corelibc/errno.h>
+#include <openenclave/corelibc/fcntl.h>
+#include <openenclave/corelibc/setjmp.h>
+#include <openenclave/corelibc/signal.h>
+#include <openenclave/corelibc/stdarg.h>
+#include <openenclave/corelibc/stdio.h>
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/corelibc/string.h>
+#include <openenclave/corelibc/sys/eventfd.h>
+#include <openenclave/corelibc/sys/ioctl.h>
+#include <openenclave/corelibc/sys/poll.h>
+#include <openenclave/corelibc/sys/select.h>
+#include <openenclave/corelibc/sys/socket.h>
+#include <openenclave/corelibc/sys/stat.h>
+#include <openenclave/corelibc/sys/syscall.h>
+#include <openenclave/corelibc/sys/uio.h>
+#include <openenclave/corelibc/sys/utsname.h>
+#include <openenclave/corelibc/unistd.h>
+#include <openenclave/internal/posix/device.h>
+#include <openenclave/internal/posix/raise.h>
+#include <openenclave/internal/print.h>
+#include <openenclave/internal/trace.h>
+
+typedef int (*ioctl_proc)(
+    int fd,
+    unsigned long request,
+    long arg1,
+    long arg2,
+    long arg3,
+    long arg4);
+
+static long _syscall(
+    long num,
+    long arg1,
+    long arg2,
+    long arg3,
+    long arg4,
+    long arg5,
+    long arg6)
+{
+    long ret = -1;
+    oe_errno = 0;
+
+    /* Handle the software system call. */
+    switch (num)
+    {
+        case OE_SYS_creat:
+        {
+            const char* pathname = (const char*)arg1;
+            oe_mode_t mode = (oe_mode_t)arg2;
+            int flags = (OE_O_CREAT | OE_O_WRONLY | OE_O_TRUNC);
+
+            ret = oe_open(pathname, flags, mode);
+            if (oe_errno == OE_ENOENT)
+            {
+                /* Not handled. Let caller dispatch this syscall. */
+                oe_errno = OE_ENOSYS;
+                goto done;
+            }
+
+            goto done;
+        }
+        case OE_SYS_open:
```

Comment:
I did a quick experiment of compiling the following code in both x64 and ARM:

```c
#include <sys/syscall.h>
#include <unistd.h>


int foo(int fd, long long offset, long count)
{
    syscall(SYS_open, fd, 0,
                   (unsigned int) (offset & 0xFFFFFFFF),
                   (unsigned int) (offset >> 32),
                   count);
}

```

When compiling for [x64](https://godbolt.org/z/VCPnDF), Sys_open has value 2:
```asm
foo:
  movq %rsi, %r8
  subq $8, %rsp
  movq %rsi, %rcx
  movq %rdx, %r9
  movl %edi, %esi
  shrq $32, %r8
  xorl %edx, %edx
  movl $2, %edi  <- Sys_open
  xorl %eax, %eax
  call syscall
  addq $8, %rsp
  ret
```

When compiling for [ARM](https://godbolt.org/z/J9xCzH), Sys_open has value 5
```asm
foo:
  str lr, [sp, #-4]!
  sub sp, sp, #12
  ldr ip, [sp, #16]
  str r3, [sp]
  mov r1, r0
  mov r3, r2
  str ip, [sp, #4]
  mov r2, #0
  mov r0, #5 <- Sys_open
  bl syscall
  add sp, sp, #12
  ldr pc, [sp], #4
```

How robust are our Sys_call enumeration values?
@gupta-ak Can you take a look at this since you added the SDK compilation support for ARM.

Reply:
Probably good to add support for arm now. I added headers for both X86_64 and ARM:

```
#if defined(__arm__)
#include <openenclave/corelibc/sys/bits/syscall_arm.h>
#elif defined(__x86_64__)
#include <openenclave/corelibc/sys/bits/syscall_x86_64.h>
#else
#error "unsupported platform"
#endif
```

No way to test it though, until we are fully building on ARM.